ext/lists.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ext
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+)
+
+func Lists() cel.EnvOption {
+	return cel.Lib(listsLib{})
+}
+
+type listsLib struct{}
+
+// LibraryName implements the SingletonLibrary interface method.
+func (listsLib) LibraryName() string {
+	return "cel.lib.ext.lists"
+}
+
+// CompileOptions implements the Library interface method.
+func (listsLib) CompileOptions() []cel.EnvOption {
+	listType := cel.ListType(cel.TypeParamType("T"))
+	return []cel.EnvOption{
+		cel.Function("slice",
+			cel.MemberOverload("list_slice",
+				[]*cel.Type{listType, cel.IntType, cel.IntType}, listType,
+				cel.FunctionBinding(func(args ...ref.Val) ref.Val {
+					list := args[0].(traits.Lister)
+					start := args[1].(types.Int)
+					end := args[2].(types.Int)
+					return listOrError(slice(list, int64(start), int64(end)))
+				}),
+			),
+		),
+	}
+}
+
+// ProgramOptions implements the Library interface method.
+func (listsLib) ProgramOptions() []cel.ProgramOption {
+	return []cel.ProgramOption{}
+}
+
+func slice(list traits.Lister, start, end int64) ([]any, error) {
+	listLength := list.Size().(types.Int)
+	if int64(listLength) < end {
+		return nil, fmt.Errorf("cannot slice(%d, %d), list is length %d", start, end, listLength)
+	}
+	if start > end {
+		return nil, fmt.Errorf("cannot slice(%d, %d), start index must be less than end index", start, end)
+	}
+
+	var newList []any
+	for i := types.Int(start); int64(i) < end; i++ {
+		newList = append(newList, list.Get(i).Value())
```

Comment:
It's possible that `i < 0`. Please check whether the result of the list `Get` call is `types.IsError` and return it as the result

Reply:
I can check `i < 0` at the top of the function, do I still need the error check on `Get()` then? If so: do I return it  as `ref.Val` or `error`?

Reply:
It's probably fine to just check for `i < 0` at the beginning, but the general contract for `Get` is that it will perform the same array index range checks as you're doing here. 

You could do any of the following:
1. Check up front (knowing it'll get checked later as well)
1. Switch to an iterator which skips the elements `< start` and stops iterating after `end - start` iterations (cleanest)
1. Check the result of `Get` to see if its a `types.IsError` and early return

I think a combination of 1 & 2 is probably best since the iterator could be optimized under the covers later.

Reply:
Okay, I'll take a look at that, I didn't realize `Get` ended up doing those checks. I started with an iterator, but ended up switching to this approach since I could put the iteration conditions into the `for`.